File: Minimum Cost Tree From Leaf Values/Leetcode_1130.cpp
class Solution {
public:
    int mctFromLeafValues(vector<int>& vals) {
        int ans = 0, N = vals.size();
        vector<int> stack = {INT_MAX};
        for (int v: vals){
            while (stack.back() <= v) {
                int v2 = stack.back();
                stack.pop_back();
                ans += v2 *min(stack.back(), v);
            }
            stack.push_back(v);
        }
        for (int i=2; i < stack.size(); ++i){
            ans += stack[i] * stack[i-1];
        }
        return ans;
    }
};

File: Minimum Cost Tree From Leaf Values/Leetcode_1130.py
class Solution:
    def mctFromLeafValues(self, vals: List[int]) -> int:
        ans, N = 0, len(vals)
        stack = [float('inf')]
        for v in vals:
            while stack[-1] <= v:
                v2 = stack.pop()
                ans += v2 * min(stack[-1], v)
            stack += [v]
        while len(stack) > 2:
            ans += stack.pop() * stack[-1]
        return ans
